lib/WUI/nhttp/req_parser.cpp

```diff
@@ -0,0 +1,177 @@
+#include "req_parser.h"
+#include "handler.h"
+#include "server.h"
+
+#include <http_req_automaton.h>
+
+#include <cstring>
+
+using namespace automata;
+using nhttp::parser::request::Names;
+using std::get;
+using std::get_if;
+using std::holds_alternative;
+using std::monostate;
+using std::string_view;
+
+namespace {
+
+const Automaton http_request(nhttp::parser::request::paths, nhttp::parser::request::transitions, nhttp::parser::request::states);
+
+}
+
+namespace nhttp::handler {
+
+RequestParser::RequestParser(const Server &server)
+    : Execution(&http_request)
+    , server(&server) {}
+
+ExecutionControl RequestParser::event(Event event) {
+    switch (event.leaving_state) {
+    case Names::MethodGet:
+        method = Method::Get;
+        return ExecutionControl::Continue;
+    case Names::MethodHead:
+        method = Method::Head;
+        return ExecutionControl::Continue;
+    case Names::MethodPost:
+        method = Method::Post;
+        return ExecutionControl::Continue;
+    case Names::MethodDelete:
+        method = Method::Delete;
+        return ExecutionControl::Continue;
+    case Names::MethodPut:
+        method = Method::Put;
+        return ExecutionControl::Continue;
+    case Names::MethodUnknown:
+        method = Method::Unknown;
+        return ExecutionControl::Continue;
+    }
+
+    switch (event.entering_state) {
+    case Names::Url:
+        if (url_size < url.size()) {
+            url[url_size++] = event.payload;
+            return ExecutionControl::Continue;
+        } else {
+            error_code = Status::UriTooLong;
+            return ExecutionControl::Continue;
+        }
+    case Names::Boundary:
+        if (boundary_size < boundary_buff.size()) {
+            boundary_buff[boundary_size++] = event.payload;
+            return ExecutionControl::Continue;
+        } else {
+            error_code = Status::RequestHeaderFieldsTooLarge;
+            return ExecutionControl::Continue;
+        }
+    case Names::XApiKey:
+        // No API key -> can't authenticate.
+        if (!api_key) {
+            auth_status = false;
+        }
+        // The first character, just starting.
+        if (holds_alternative<monostate>(auth_status)) {
+            auth_status = static_cast<uint8_t>(0);
+        }
+        if (uint8_t *idx = get_if<uint8_t>(&auth_status); idx != nullptr) {
+            const size_t key_len = strlen(api_key);
```

Comment:
Is it possible for the `api_key` to be `nullptr` in some cases? You have a check for it above, but that doesn't protect this line from being executed.

Reply:
- When `!api_key`, it sets auth_status to some `bool`. 
- The line 75 won't get executed, as the condition on line 74 isn't true
- `get_if` returns `nullptr`, as `auth_status` still holds `bool` from the first step. Therefore, this line is protected from such condition.

It is getting late and I am starting to be desperate. What am I missing? 😭

Reply:
TBH, clang analyzer pointed this out, not sure if such a scenario can be real:
```
req_parser.cpp:30:5: 'Default' branch taken. Execution continues on line 51
req_parser.cpp:51:5: Control jumps to 'case XApiKey:'  at line 68
req_parser.cpp:70:13: Assuming field 'api_key' is null
req_parser.cpp:70:9: Taking true branch
req_parser.cpp:74:13: Assuming the condition is false
req_parser.cpp:74:9: Taking false branch
req_parser.cpp:77:9: Taking true branch
req_parser.cpp:78:36: Null pointer passed as 1st argument to string length function
```

Reply:
So, Alan is right that it can't (couldn't) happen. His analysis is, however, more complex than what clang analyzer can do ‒ he takes how it behaved in the previous loop into account, while the clang analyzer doesn't reason in such a complex way.

Nevertheless, the fact that one needs to do such complex analysis to prove it can't happen is not really reader-friendly, so the best way is to just throw another `return ...` statement in there just to make sure.